src/main/java/bgu/spl/net/impl/stomp/LibraryProtocol.java

```diff
@@ -13,59 +13,168 @@
     private Client _client;
     private int _connectionId;
     private Connections<Frame> _connections;
+    private Integer _messagesSent;
 
     @Override
     public void start(int connectionId, Connections<Frame> connections) {
         _connectionId = connectionId;
         _connections = connections;
+        _messagesSent = 0;
     }
 
     @Override
     public void process(Frame msg) {
         switch (msg.getCommand()) {
             case CONNECT:
-                _connections.send(_connectionId, handleConnect(msg));
+                handleConnect(msg);
                 break;
             case SEND:
                 handleSend(msg);
                 break;
-            // TODO: Add the rest of the handling. For all of the commands
+            case SUBSCRIBE:
+                handleSubscribe(msg);
+                break;
+            case UNSUBSCRIBE:
+                handleUnsubscribe(msg);
+                break;
+            case DISCONNECT:
+                handleDisconnect(msg);
+                break;
         }
     }
 
+    private void handleDisconnect(Frame msg) {
+        sendReceipt(msg);
+        _connections.disconnect(_connectionId);
+        _client.connected = false;
+        _shouldTerminate = true;
+    }
+
     /**
-     *  Adds a receipt id needed according to {@code msg}'s headers
+     * This function checks for the existence of the header {@code header}, if it doesn't exist the function sends the
+     * proper error Frame
+     *
+     * @param msg    The Frame to check for the header
+     * @param header The header to be checked for existence
+     * @return Does the header exist
+     */
+    private boolean checkForHeader(Frame msg, String header, String info) {
+        if (msg.getHeader(header) == null) {
+            Frame f = createErrorFrame("MalFormed Frame - missing " + header + " header");
+            if (info != null) addErrorBody(msg, f, info);
+            _connections.send(_connectionId,
+                    f);
+            return false;
+        }
+        return true;
+    }
+
+    private boolean checkForHeader(Frame msg, String header) {
+        return checkForHeader(msg, header, null);
+    }
+
+    private void handleSubscribe(Frame msg) {
+        if (!checkForHeader(msg, "destination")) return;
+        if (!checkForHeader(msg, "id")) return;
+        _connections.subscribe(msg.getHeader("destination"), _connectionId, Integer.parseInt(msg.getHeader("id")));
+        sendReceipt(msg);
+    }
+
+    private void handleUnsubscribe(Frame msg) {
+        if (!checkForHeader(msg, "id")) return;
+        try {
+            _connections.unsubscribe(msg.getHeader("destination"), _connectionId, Integer.parseInt(msg.getHeader("id")));
+        }
+        // We ignore a NullPointer exception because one will be thrown if the subscriptionId doesn't exist
+        catch (NullPointerException ignored) {
+        }
+        sendReceipt(msg);
```

Comment:
I'm not sure that's the right handle...

Reply:
I suppose you're right, I will change it

Reply:
i meant, you should send an Error, and not Receipt in this case... :sweat_smile:

Reply:
The receipt is sent anyway, since the handling for this situation is supposed to be that nothing happens. Should I change that?

Reply:
I think so. because receipt is like confirming the action was made. in this case, an Error with "not registered to this genre" or something like that should be sent

Reply:
Done